File: src-tauri/src/services/browser_error.rs
use serde::{Deserialize, Serialize};
use std::fmt;

/// Structured error types for browser operations
/// This ensures type-safe error handling between Rust and TypeScript
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "code", content = "context")]
pub enum BrowserError {
    /// Session with the given ID was not found
    #[serde(rename = "SESSION_NOT_FOUND")]
    SessionNotFound { session_id: String },

    /// Session was already closed
    #[serde(rename = "SESSION_CLOSED")]
    SessionClosed { session_id: String },

    /// Browser window was not found
    #[serde(rename = "WINDOW_NOT_FOUND")]
    WindowNotFound { session_id: String },

    /// Element could not be found with the given selector
    #[serde(rename = "ELEMENT_NOT_FOUND")]
    ElementNotFound {
        selector: String,
        session_id: String,
    },

    /// Element exists but is not interactable
    #[serde(rename = "ELEMENT_NOT_INTERACTABLE")]
    ElementNotInteractable {
        selector: String,
        reason: String,
        session_id: String,
    },

    /// Navigation failed
    #[serde(rename = "NAVIGATION_FAILED")]
    NavigationFailed {
        url: String,
        reason: String,
        session_id: String,
    },

    /// Script execution failed
    #[serde(rename = "SCRIPT_EXECUTION_FAILED")]
    ScriptExecutionFailed { reason: String, session_id: String },

    /// Operation timed out
    #[serde(rename = "TIMEOUT")]
    Timeout {
        operation: String,
        duration_ms: u64,
        session_id: String,
    },

    /// Lock acquisition failed
    #[serde(rename = "LOCK_FAILED")]
    LockFailed { reason: String },

    /// Invalid parameter
    #[serde(rename = "INVALID_PARAMETER")]
    InvalidParameter { parameter: String, reason: String },

    /// Generic error
    #[serde(rename = "UNKNOWN")]
    Unknown { message: String },
}

impl BrowserError {
    /// Convert error to user-friendly message
    pub fn to_message(&self) -> String {
        match self {
            BrowserError::SessionNotFound { session_id } => {
                format!("Session '{session_id}' not found")
            }
            BrowserError::SessionClosed { session_id } => {
                format!("Session '{session_id}' was already closed")
            }
            BrowserError::WindowNotFound { session_id } => {
                format!("Browser window for session '{session_id}' not found")
            }
            BrowserError::ElementNotFound {
                selector,
                session_id,
            } => {
                format!("Element with selector '{selector}' not found in session '{session_id}'")
            }
            BrowserError::ElementNotInteractable {
                selector,
                reason,
                session_id,
            } => {
                format!(
                    "Element '{selector}' in session '{session_id}' is not interactable: {reason}"
                )
            }
            BrowserError::NavigationFailed {
                url,
                reason,
                session_id,
            } => {
                format!("Navigation to '{url}' failed in session '{session_id}': {reason}")
            }
            BrowserError::ScriptExecutionFailed { reason, session_id } => {
                format!("Script execution failed in session '{session_id}': {reason}")
            }
            BrowserError::Timeout {
                operation,
                duration_ms,
                session_id,
            } => {
                format!(
                    "Operation '{operation}' timed out after {duration_ms}ms in session '{session_id}'"
                )
            }
            BrowserError::LockFailed { reason } => {
                format!("Failed to acquire lock: {reason}")
            }
            BrowserError::InvalidParameter { parameter, reason } => {
                format!("Invalid parameter '{parameter}': {reason}")
            }
            BrowserError::Unknown { message } => message.clone(),
        }
    }

    /// Get error code as string
    pub fn code(&self) -> &'static str {
        match self {
            BrowserError::SessionNotFound { .. } => "SESSION_NOT_FOUND",
            BrowserError::SessionClosed { .. } => "SESSION_CLOSED",
            BrowserError::WindowNotFound { .. } => "WINDOW_NOT_FOUND",
            BrowserError::ElementNotFound { .. } => "ELEMENT_NOT_FOUND",
            BrowserError::ElementNotInteractable { .. } => "ELEMENT_NOT_INTERACTABLE",
            BrowserError::NavigationFailed { .. } => "NAVIGATION_FAILED",
            BrowserError::ScriptExecutionFailed { .. } => "SCRIPT_EXECUTION_FAILED",
            BrowserError::Timeout { .. } => "TIMEOUT",
            BrowserError::LockFailed { .. } => "LOCK_FAILED",
            BrowserError::InvalidParameter { .. } => "INVALID_PARAMETER",
            BrowserError::Unknown { .. } => "UNKNOWN",
        }
    }
}

impl fmt::Display for BrowserError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.to_message())
    }
}

impl From<BrowserError> for String {
    fn from(error: BrowserError) -> Self {
        // Serialize to JSON for structured error passing
        serde_json::to_string(&error).unwrap_or_else(|_| error.to_message())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_error_serialization() {
        let error = BrowserError::SessionNotFound {
            session_id: "test-123".to_string(),
        };

        let json = serde_json::to_string(&error).unwrap();
        assert!(json.contains("SESSION_NOT_FOUND"));
        assert!(json.contains("test-123"));

        let deserialized: BrowserError = serde_json::from_str(&json).unwrap();
        assert_eq!(error.code(), deserialized.code());
    }

    #[test]
    fn test_error_messages() {
        let error = BrowserError::ElementNotFound {
            selector: ".button".to_string(),
            session_id: "sess-1".to_string(),
        };

        assert!(error.to_message().contains(".button"));
        assert!(error.to_message().contains("sess-1"));
    }
}

File: src/features/tools/browser-tools/browser-error.ts
/**
 * Structured browser error types
 * These types match the Rust BrowserError enum for type-safe error handling
 */

/**
 * Browser error codes
 */
export enum BrowserErrorCode {
  SESSION_NOT_FOUND = 'SESSION_NOT_FOUND',
  SESSION_CLOSED = 'SESSION_CLOSED',
  WINDOW_NOT_FOUND = 'WINDOW_NOT_FOUND',
  ELEMENT_NOT_FOUND = 'ELEMENT_NOT_FOUND',
  ELEMENT_NOT_INTERACTABLE = 'ELEMENT_NOT_INTERACTABLE',
  NAVIGATION_FAILED = 'NAVIGATION_FAILED',
  SCRIPT_EXECUTION_FAILED = 'SCRIPT_EXECUTION_FAILED',
  TIMEOUT = 'TIMEOUT',
  LOCK_FAILED = 'LOCK_FAILED',
  INVALID_PARAMETER = 'INVALID_PARAMETER',
  UNKNOWN = 'UNKNOWN',
}

/**
 * Base interface for structured browser errors
 */
interface BaseBrowserError {
  code: BrowserErrorCode;
}

/**
 * Session not found error
 */
export interface SessionNotFoundError extends BaseBrowserError {
  code: BrowserErrorCode.SESSION_NOT_FOUND;
  context: {
    session_id: string;
  };
}

/**
 * Session closed error
 */
export interface SessionClosedError extends BaseBrowserError {
  code: BrowserErrorCode.SESSION_CLOSED;
  context: {
    session_id: string;
  };
}

/**
 * Window not found error
 */
export interface WindowNotFoundError extends BaseBrowserError {
  code: BrowserErrorCode.WINDOW_NOT_FOUND;
  context: {
    session_id: string;
  };
}

/**
 * Element not found error
 */
export interface ElementNotFoundError extends BaseBrowserError {
  code: BrowserErrorCode.ELEMENT_NOT_FOUND;
  context: {
    selector: string;
    session_id: string;
  };
}

/**
 * Element not interactable error
 */
export interface ElementNotInteractableError extends BaseBrowserError {
  code: BrowserErrorCode.ELEMENT_NOT_INTERACTABLE;
  context: {
    selector: string;
    reason: string;
    session_id: string;
  };
}

/**
 * Navigation failed error
 */
export interface NavigationFailedError extends BaseBrowserError {
  code: BrowserErrorCode.NAVIGATION_FAILED;
  context: {
    url: string;
    reason: string;
    session_id: string;
  };
}

/**
 * Script execution failed error
 */
export interface ScriptExecutionFailedError extends BaseBrowserError {
  code: BrowserErrorCode.SCRIPT_EXECUTION_FAILED;
  context: {
    reason: string;
    session_id: string;
  };
}

/**
 * Timeout error
 */
export interface TimeoutError extends BaseBrowserError {
  code: BrowserErrorCode.TIMEOUT;
  context: {
    operation: string;
    duration_ms: number;
    session_id: string;
  };
}

/**
 * Lock acquisition failed error
 */
export interface LockFailedError extends BaseBrowserError {
  code: BrowserErrorCode.LOCK_FAILED;
  context: {
    reason: string;
  };
}

/**
 * Invalid parameter error
 */
export interface InvalidParameterError extends BaseBrowserError {
  code: BrowserErrorCode.INVALID_PARAMETER;
  context: {
    parameter: string;
    reason: string;
  };
}

/**
 * Unknown error
 */
export interface UnknownError extends BaseBrowserError {
  code: BrowserErrorCode.UNKNOWN;
  context: {
    message: string;
  };
}

/**
 * Union type of all browser errors
 */
export type BrowserError =
  | SessionNotFoundError
  | SessionClosedError
  | WindowNotFoundError
  | ElementNotFoundError
  | ElementNotInteractableError
  | NavigationFailedError
  | ScriptExecutionFailedError
  | TimeoutError
  | LockFailedError
  | InvalidParameterError
  | UnknownError;

/**
 * Type guard to check if an error is a structured browser error
 */
export function isBrowserError(error: unknown): error is BrowserError {
  if (typeof error !== 'object' || error === null) {
    return false;
  }

  const err = error as Record<string, unknown>;
  return (
    typeof err.code === 'string' &&
    Object.values(BrowserErrorCode).includes(err.code as BrowserErrorCode) &&
    typeof err.context === 'object' &&
    err.context !== null
  );
}

/**
 * Parse error from backend response
 * Tries to parse as structured error, falls back to string error
 */
export function parseBrowserError(error: unknown): BrowserError | string {
  // Try to parse as JSON string first
  if (typeof error === 'string') {
    try {
      const parsed = JSON.parse(error);
      if (isBrowserError(parsed)) {
        return parsed;
      }
    } catch {
      // Not JSON, return as string
      return error;
    }
  }

  // Check if it's already a structured error
  if (isBrowserError(error)) {
    return error;
  }

  // Handle Error objects
  if (error instanceof Error) {
    try {
      const parsed = JSON.parse(error.message);
      if (isBrowserError(parsed)) {
        return parsed;
      }
    } catch {
      return error.message;
    }
  }

  // Fallback to string
  return String(error);
}

/**
 * Get user-friendly error message from browser error
 */
export function getBrowserErrorMessage(error: BrowserError): string {
  switch (error.code) {
    case BrowserErrorCode.SESSION_NOT_FOUND:
      return `Session '${error.context.session_id}' not found`;
    case BrowserErrorCode.SESSION_CLOSED:
      return `Session '${error.context.session_id}' was already closed`;
    case BrowserErrorCode.WINDOW_NOT_FOUND:
      return `Browser window for session '${error.context.session_id}' not found`;
    case BrowserErrorCode.ELEMENT_NOT_FOUND:
      return `Element with selector '${error.context.selector}' not found in session '${error.context.session_id}'`;
    case BrowserErrorCode.ELEMENT_NOT_INTERACTABLE:
      return `Element '${error.context.selector}' in session '${error.context.session_id}' is not interactable: ${error.context.reason}`;
    case BrowserErrorCode.NAVIGATION_FAILED:
      return `Navigation to '${error.context.url}' failed in session '${error.context.session_id}': ${error.context.reason}`;
    case BrowserErrorCode.SCRIPT_EXECUTION_FAILED:
      return `Script execution failed in session '${error.context.session_id}': ${error.context.reason}`;
    case BrowserErrorCode.TIMEOUT:
      return `Operation '${error.context.operation}' timed out after ${error.context.duration_ms}ms in session '${error.context.session_id}'`;
    case BrowserErrorCode.LOCK_FAILED:
      return `Failed to acquire lock: ${error.context.reason}`;
    case BrowserErrorCode.INVALID_PARAMETER:
      return `Invalid parameter '${error.context.parameter}': ${error.context.reason}`;
    case BrowserErrorCode.UNKNOWN:
      return error.context.message;
  }
}
